### PR TITLE
JEP371 ClassData Support and JEP383 Support (Part 3)

### PR DESCRIPTION
**Pull request by babsingh**
**JEP371 ClassData Support:**
1. Implement `Class.setClassData(...) and Class.getClassData()`
2. Store `classData` before init in `ClassLoader_defineClassImpl1`
3. Implement `Access.classData(...)`
4. Implement `MethodHandleNatives.classData(...)`
5. Implement `ClassDefiner.defineClass(initOption, classData)`
6. Implement `Lookup.defineHiddenClassWithClassData(...)`
7. Implement `MethodHandles.classData(...)`

**JEP383 Support (Part 3):**
1. Implement `MethodHandles.permuteArgumentChecks(...)`
2. Implement `MethodHandles.collectReturnValue(...)`
3. Implement `VarHandle.target()`
4. Implement `VarHandle.asDirect()`
5. Implement `VarHandle.isDirect()`
6. Implement `VarHandle.getMethodHandle(...)`

Related: https://github.com/eclipse/openj9/issues/9625

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
This PR shouldn't be merged. Yesterday, I had to modify my changes to work with @hangshao0's JEP371 changes while rebasing: https://github.com/eclipse/openj9/issues/9328. Last night, I have noticed functional issues with the new JEP371 changes. The previous usage of `Unsafe` in `classDefine` worked with JEP383 but the new JEP371 changes throws an exception during class initialization. I will document the failure and report details in an issue on Monday.

**Comment by babsingh** [comment]
Updated my changes to work with JEP371. The new changes are added in the HEAD commit. JEP383 depends on JEP371. This PR contains code which extends JEP371's functionality. @hangshao0 can you please peer-review the new code related to JEP371?

**Comment by babsingh** [comment]
Documenting the error when `Class.classData` is set after `ClassLoader_defineClassImpl1`, which defines and initializes the class:
```
WARNING: Using incubator modules: jdk.incubator.foreign
Exception in thread "main" java.lang.AssertionError: java.lang.ExceptionInInitializerError
	at java.base/java.lang.invoke.MemoryAccessVarHandleGenerator.generateHandleFactory(MemoryAccessVarHandleGenerator.java:173)
	at java.base/java.lang.invoke.VarHandles.lambda$makeMemoryAddressViewHandle$0(VarHandles.java:331)
	at java.base/java.lang.invoke.VarHandles$$Lambda$15/00000000D84C41A0.apply(Unknown Source)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1708)
	at java.base/java.lang.invoke.VarHandles.makeMemoryAddressViewHandle(VarHandles.java:329)
	at java.base/java.lang.invoke.MethodHandleImpl$1.memoryAccessVarHandle(MethodHandleImpl.java:1807)
	at jdk.incubator.foreign/jdk.incubator.foreign.MemoryHandles.varHandle(MemoryHandles.java:235)
	at jdk.incubator.foreign/jdk.incubator.foreign.MemoryHandles.varHandle(MemoryHandles.java:204)
	at TestMemoryScope.main(TestMemoryScope.java:17)
Caused by: java.lang.ExceptionInInitializerError
	at java.base/java.lang.J9VMInternals.ensureError(J9VMInternals.java:186)
	at java.base/java.lang.J9VMInternals.recordInitializationFailure(J9VMInternals.java:175)
	at java.base/java.lang.ClassLoader.defineClassImpl1(Native Method)
	at java.base/java.lang.ClassLoader.defineClassInternal(ClassLoader.java:515)
	at java.base/java.lang.Access.defineClass(Access.java:336)
	at java.base/java.lang.invoke.MethodHandles$Lookup$ClassDefiner$2.run(MethodHandles.java:2011)
	at java.base/java.lang.invoke.MethodHandles$Lookup$ClassDefiner$2.run(MethodHandles.java:2006)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:678)
	at java.base/java.lang.invoke.MethodHandles$Lookup$ClassDefiner.defineClass(MethodHandles.java:2006)
	at java.base/java.lang.invoke.MethodHandles$Lookup.defineHiddenClassWithClassData(MethodHandles.java:2046)
	at java.base/java.lang.invoke.MemoryAccessVarHandleGenerator.generateHandleFactory(MemoryAccessVarHandleGenerator.java:160)
	... 8 more
Caused by: java.lang.NullPointerException
	at java.base/java.lang.invoke.MemoryAccessVarHandleByteHelper0/00000000D84E2780.<clinit>(Unknown Source)
	... 17 more
```

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -2282,45 +2282,83 @@ int toFlag() {
...
+			}
+
+			Class<?> defineClass(boolean initOption, Object classData) {
+				Class<?> ret = AccessController.doPrivileged(new PrivilegedAction<Class<?>>() {
+					@Override
+					public Class<?> run() {
+						JavaLangAccess jlAccess = SharedSecrets.getJavaLangAccess();
+						Class<?> lookupClass = lookup.lookupClass();
+						return jlAccess.defineClass(lookupClass.getClassLoader(), lookupClass, className, classBytes,
+								jlAccess.protectionDomain(lookupClass), initOption, ClassOptFlags, classData);
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2320]: This duplicates the body of `defineClass`.  Can the ` Class<?> defineClass(boolean initOption) ` version forward to this one with a null `classData`?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -2282,45 +2282,83 @@ int toFlag() {
...
 				});
 				return ret;
 			}
 		}
+
 		/**
 		 * Constructs a new hidden class from an array of class data bytes.
 		 * 
-		 * @param bytes the class data bytes of the hidden class to be defined.  
+		 * @param bytes the class data bytes of the hidden class to be defined.
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2330]: `class data bytes` may be confusing given the addition of the `classData` object

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 3]
```diff
@@ -2282,45 +2282,83 @@ int toFlag() {
...
 		 * @throws IllegalAccessException if this Lookup does not have full privilege access.
 		 */
-
 		public Lookup defineHiddenClass(byte[] bytes, boolean initOption, ClassOption... classOptions) throws IllegalAccessException {
+			ClassDefiner definer = classDefiner(bytes, classOptions);
+			return new Lookup(definer.defineClass(initOption));
+		}
+		
+		/**
+		 * Constructs a new hidden class from an array of class data bytes.
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2343]:
```suggestion
		 * Constructs a new hidden class from an array of class file bytes.
```

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 4]
```diff
@@ -2282,45 +2282,83 @@ int toFlag() {
...
+		 * 
+		 * @param bytes the class data bytes of the hidden class to be defined.
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2345]:
```suggestion
		 * @param bytes the class file bytes of the hidden class to be defined.
```

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 5]
```diff
@@ -5466,12 +5500,96 @@ public Object helper(Object[] arguments) throws Throwable {
...
+		return validatePermutationArray(permuteType, handleType, permute);
 	}
 	
-	static MethodHandle collectReturnValue(MethodHandle mh1, MethodHandle mh2) {
-		throw OpenJDKCompileStub.OpenJDKCompileStubThrowError();
+	/**
+	 * Return the classData stored in the accessClass of the Lookup object.
+	 * 
+	 * @param caller Lookup object used to verify privileged access and retrieve classData.
+	 * @param unused.
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:5523]: Why an unused parameter?

File: jcl/src/java.base/share/classes/java/lang/Class.java [hunk 6]
```diff
@@ -266,6 +266,10 @@ static Unsafe getUnsafe() {
 	private Class<?> nestHost;
 /*[ENDIF] Java11*/
 	
+/*[IF Java15]*/
+	private transient Object classData;
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/Class.java:270]: Is this field visible to Reflection in the RI?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 7]
```diff
@@ -2282,45 +2282,83 @@ int toFlag() {
...
+		 * @param bytes the class data bytes of the hidden class to be defined.
+		 * @param classData the classData to be stored in the hidden class.
+		 * @param classOptions the {@link ClassOption} to define the hidden class.
+		 * 
+		 * @return A Lookup object of the newly created hidden class.
+		 * @throws IllegalAccessException if this Lookup does not have full privilege access.
+		 */
+		Lookup defineHiddenClassWithClassData(byte[] bytes, Object classData, ClassOption... classOptions) throws IllegalAccessException {
+			ClassDefiner definer = classDefiner(bytes, classOptions);
+			return new Lookup(definer.defineClass(true, classData));
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2354]: How did we determine that `true` aka initialize is the right value here?

Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2354]: From the [documentation](https://github.com/ibmruntimes/openj9-openjdk-jdk15/blob/88d915c8537bf7e6a9828c0523cb7478cf20ed10/src/java.base/share/classes/java/lang/invoke/MethodHandles.java#L1973-L1974). `defineHiddenClassWithClassData` is a package private method so its documentation does not get published in the official docs.

**Comment by babsingh** [comment]
> Can you also rename this PR? From the contents, it's mostly for adding `classData` support but hard to tell that from the title

Renamed the PR, and distinguished between JEP371 and JEP383 work in the top description.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 1 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:2311]: Yes. Addressed in the HEAD commit.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 2 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Reworded `class data bytes` to `class file bytes` in the HEAD commit.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 3 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Incorporated in the HEAD commit.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 4 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Incorporated in the HEAD commit.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java (hunk 5 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java:5515]: RI introduced `MethodHandles.classData` with an `unused String`. We had to match the RI in order to support calls from the JEP383 API.

File: jcl/src/java.base/share/classes/java/lang/Class.java (hunk 6 above)
Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/Class.java:270]: The `Class.classData` field is listed in [Reflection.fieldFilterMap](https://github.com/ibmruntimes/openj9-openjdk-jdk15/blob/88d915c8537bf7e6a9828c0523cb7478cf20ed10/src/java.base/share/classes/jdk/internal/reflect/Reflection.java#L55). It gets filtered out from public-view. I don't think it is visible to Reflection in the RI. Assuming, I have correctly read the code.

**Comment by babsingh** [comment]
@DanHeidinga All comments have been addressed, and the suggested changes are included in the HEAD commit.

**Comment by DanHeidinga** [comment]
Jenkins test sanity zlinux jdk15,jdk11

**Pull request merged**